libass/ass_bitmap.c

```diff
@@ -146,101 +146,81 @@ static bool alloc_bitmap_buffer(const BitmapEngine *engine, Bitmap *bm, int w, i
     return true;
 }
 
-Bitmap *alloc_bitmap(const BitmapEngine *engine, int w, int h, bool zero)
-{
-    Bitmap *bm = malloc(sizeof(Bitmap));
-    if (!bm)
-        return NULL;
-    if (!alloc_bitmap_buffer(engine, bm, w, h, zero)) {
-        free(bm);
-        return NULL;
-    }
-    bm->left = bm->top = 0;
-    return bm;
-}
-
 bool realloc_bitmap(const BitmapEngine *engine, Bitmap *bm, int w, int h)
 {
     uint8_t *old = bm->buffer;
-    if (!alloc_bitmap_buffer(engine, bm, w, h, false))
+    if (!alloc_bitmap(engine, bm, w, h, false))
         return false;
     ass_aligned_free(old);
     return true;
 }
 
 void ass_free_bitmap(Bitmap *bm)
 {
-    if (bm)
-        ass_aligned_free(bm->buffer);
-    free(bm);
+    ass_aligned_free(bm->buffer);
 }
 
-Bitmap *copy_bitmap(const BitmapEngine *engine, const Bitmap *src)
+bool copy_bitmap(const BitmapEngine *engine, Bitmap *dst, const Bitmap *src)
 {
-    Bitmap *dst = alloc_bitmap(engine, src->w, src->h, false);
-    if (!dst)
-        return NULL;
+    if (!src->buffer) {
+        memset(dst, 0, sizeof(*dst));
+        return true;
+    }
+    if (!alloc_bitmap(engine, dst, src->w, src->h, false))
+        return false;
     dst->left = src->left;
-    dst->top = src->top;
+    dst->top  = src->top;
     memcpy(dst->buffer, src->buffer, src->stride * src->h);
-    return dst;
+    return true;
 }
 
-Bitmap *outline_to_bitmap(ASS_Renderer *render_priv,
-                          ASS_Outline *outline1, ASS_Outline *outline2,
-                          int bord)
+bool outline_to_bitmap(ASS_Renderer *render_priv, Bitmap *bm,
+                       ASS_Outline *outline1, ASS_Outline *outline2)
 {
     RasterizerData *rst = &render_priv->rasterizer;
     if (outline1 && !rasterizer_set_outline(rst, outline1, false)) {
         ass_msg(render_priv->library, MSGL_WARN, "Failed to process glyph outline!\n");
-        return NULL;
+        return false;
     }
     if (outline2 && !rasterizer_set_outline(rst, outline2, !!outline1)) {
         ass_msg(render_priv->library, MSGL_WARN, "Failed to process glyph outline!\n");
-        return NULL;
+        return false;
     }
-
     if (rst->bbox.x_min > rst->bbox.x_max || rst->bbox.y_min > rst->bbox.y_max)
-        return NULL;
-
-    if (bord < 0 || bord > INT_MAX / 2)
-        return NULL;
-    if (rst->bbox.x_max > INT_MAX - 63 || rst->bbox.y_max > INT_MAX - 63)
-        return NULL;
+        return false;
 
-    int x_min = rst->bbox.x_min >> 6;
-    int y_min = rst->bbox.y_min >> 6;
-    int x_max = (rst->bbox.x_max + 63) >> 6;
-    int y_max = (rst->bbox.y_max + 63) >> 6;
-    int w = x_max - x_min;
-    int h = y_max - y_min;
+    // enlarge by 1/64th of pixel to bypass slow rasterizer path, add 1 pixel for shift_bitmap
```

Comment:
Then polyline point lies on image border or outside that triggers additional cropping pass. Here I've simply replaced 1 pixel border with minimal possible values.

Reply:
Why is the 1/64 pixel needed at all? My impression was that the only reason this code used to add a pixel on the left and top was that the margin had to be symmetric and we needed a pixel added on the right and bottom, so I expected this change to stop adding _anything_ on the left and top now, not add 1/64.

Reply:
Zero border can lead to ~1.5 slowdown in rasterization of small letters so I would like to avoid that. Well, it's quite rare case then letter is exactly at pixel border but better to be safe.

Reply:
OK.

I’d just like to understand why the slowdown happens and is unavoidable. Is it a byproduct of approximating the curve with a polyline?

Reply:
It's that: https://github.com/libass/libass/blob/master/libass/ass_rasterizer.c#L769

Then `64 * bm->left == bbox.x_min` (or for any other image border) there is a possibility that some of polyline edges are situated fully outside of the image rectangle (coincide with rectangle border). Such edges are excluded from rasterization and contribute to winding number calculation instead. Simplest method to avoid that is to offset polyline bounding rectangle by minimum possible value.